app/com/arpnetworking/metrics/portal/reports/impl/DatabaseReportExecutionRepository.java

```diff
@@ -0,0 +1,311 @@
+            newOrUpdatedExecution.setState(state);
+
+            switch (state) {
+                case STARTED:
+                    newOrUpdatedExecution.setStartedAt(Instant.now());
+                    newOrUpdatedExecution.setCompletedAt(null);
+                    break;
+                case FAILURE:
+                    if (error == null) {
+                        throw new IllegalArgumentException("Error is null for a failed execution.");
+                    }
+                    newOrUpdatedExecution.setError(new ReportExecution.ErrorString(Throwables.getStackTraceAsString(error)));
+                    newOrUpdatedExecution.setCompletedAt(Instant.now());
+                    break;
+                case SUCCESS:
+                    if (result == null) {
+                        throw new IllegalArgumentException("Result is null for a successful execution.");
+                    }
+                    newOrUpdatedExecution.setResult(result);
+                    newOrUpdatedExecution.setCompletedAt(Instant.now());
+                    break;
+                default:
+                    throw new AssertionError("unexpected state: " + state);
+            }
+
+            if (existingExecution.isPresent()) {
+                _ebeanServer.update(newOrUpdatedExecution);
+            } else {
+                _ebeanServer.save(newOrUpdatedExecution);
+            }
+
+            LOGGER.debug()
+                    .setMessage("Updated report execution")
+                    .addData("report.uuid", reportId)
+                    .addData("scheduled", scheduled)
+                    .addData("state", state)
+                    .log();
+            transaction.commit();
+            // CHECKSTYLE.OFF: IllegalCatchCheck
+        } catch (final RuntimeException e) {
+            // CHECKSTYLE.ON: IllegalCatchCheck
+            LOGGER.error()
+                    .setMessage("Failed to update report executions")
+                    .addData("report.uuid", reportId)
+                    .addData("scheduled", scheduled)
+                    .addData("state", state)
+                    .setThrowable(e)
+                    .log();
+            throw new PersistenceException("Failed to update report executions", e);
+        }
+    }
+
+    private JobExecution<Report.Result> toInternalModel(final ReportExecution beanModel) {
+        final ReportExecution.State state = beanModel.getState();
+        switch (state) {
+            case STARTED:
+                return new JobExecution.Started.Builder<Report.Result>()
+                        .setJobId(beanModel.getReport().getUuid())
+                        .setScheduled(beanModel.getScheduled())
+                        .setStartedAt(beanModel.getStartedAt())
+                        .build();
+            case FAILURE:
+                @Nullable final Throwable throwable = beanModel.getError() == null ? null : beanModel.getError().getThrowable();
```

Comment:
Above it seems that error is required (when transitioning to failure), but here it seems optional. Did I miss something?

Reply:
(is this constraint represented in the model?)

Reply:
It's nullable here because it is nullable in the data model. The constraint is validated in the internal model's builder.

Reply:
Yikes. We have a stricter internal model than data model? I guess it's not dangerous, just odd.

Reply:
I don't think it would be safe to mark the fields as NonNull anyway since they're essentially coming from the outside world. I consider it defensive at an IO boundary. There's nothing stopping someone from populating a row with a nonsensical state from the command line, for example.